File: cakk-api/src/main/java/com/cakk/api/service/user/SignService.java
package com.cakk.api.service.user;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import lombok.RequiredArgsConstructor;

import com.cakk.api.dto.request.user.UserSignInRequest;
import com.cakk.api.dto.request.user.UserSignUpRequest;
import com.cakk.api.dto.response.user.JwtResponse;
import com.cakk.api.factory.OidcProviderFactory;
import com.cakk.api.mapper.UserMapper;
import com.cakk.api.provider.jwt.JwtProvider;
import com.cakk.common.enums.ReturnCode;
import com.cakk.common.exception.CakkException;
import com.cakk.core.facade.user.UserManageFacade;
import com.cakk.core.facade.user.UserReadFacade;
import com.cakk.domain.mysql.entity.user.User;
import com.cakk.domain.redis.repository.TokenRedisRepository;

@Service
@RequiredArgsConstructor
public class SignService {

	private final OidcProviderFactory oidcProviderFactory;
	private final JwtProvider jwtProvider;

	private final UserReadFacade userReadFacade;
	private final UserManageFacade userManageFacade;
	private final TokenRedisRepository tokenRedisRepository;

	@Transactional
	public JwtResponse signUp(final UserSignUpRequest dto) {
		final String providerId = oidcProviderFactory.getProviderId(dto.provider(), dto.idToken());
		final User user = userManageFacade.create(UserMapper.supplyUserBy(dto, providerId));

		return JwtResponse.from(jwtProvider.generateToken(user));
	}

	@Transactional(readOnly = true)
	public JwtResponse signIn(final UserSignInRequest dto) {
		final String providerId = oidcProviderFactory.getProviderId(dto.provider(), dto.idToken());
		final User user = userReadFacade.findByProviderId(providerId);

		return JwtResponse.from(jwtProvider.generateToken(user));
	}

	public void signOut(final String accessToken, final String refreshToken) {
		final long accessTokenExpiredSecond = jwtProvider.getTokenExpiredSecond(accessToken) - System.currentTimeMillis();
		final long refreshTokenExpiredSecond = jwtProvider.getTokenExpiredSecond(refreshToken) - System.currentTimeMillis();

		tokenRedisRepository.registerBlackList(accessToken, accessTokenExpiredSecond);
		tokenRedisRepository.registerBlackList(refreshToken, refreshTokenExpiredSecond);
	}

	@Transactional(readOnly = true)
	public JwtResponse recreateToken(final String refreshToken) {
		final boolean isBlackList = tokenRedisRepository.isBlackListToken(refreshToken);

		if (isBlackList) {
			throw new CakkException(ReturnCode.BLACK_LIST_TOKEN);
		}

		final User user = jwtProvider.getUser(refreshToken);
		final long expiredSecond = jwtProvider.getTokenExpiredSecond(refreshToken) - System.currentTimeMillis();

		tokenRedisRepository.registerBlackList(refreshToken, expiredSecond);

		return JwtResponse.from(jwtProvider.generateToken(user));
	}
}

File: cakk-core/src/main/kotlin/com/cakk/core/service/user/SignService.kt
package com.cakk.core.service.user

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

import com.cakk.common.enums.ReturnCode
import com.cakk.common.exception.CakkException
import com.cakk.common.utils.currentTimeMillis
import com.cakk.core.dispatcher.OidcProviderDispatcher
import com.cakk.core.dto.param.user.UserSignInParam
import com.cakk.core.dto.param.user.UserSignUpParam
import com.cakk.core.dto.response.user.JwtResponse
import com.cakk.core.facade.user.UserManageFacade
import com.cakk.core.facade.user.UserReadFacade
import com.cakk.core.mapper.supplyJwtResponseBy
import com.cakk.core.mapper.supplyUserBy
import com.cakk.core.provider.jwt.JwtProvider
import com.cakk.infrastructure.cache.repository.TokenRedisRepository

@Service
class SignService(
    private val userReadFacade: UserReadFacade,
    private val userManageFacade: UserManageFacade,
    private val tokenRedisRepository: TokenRedisRepository,
    private val oidcProviderDispatcher: OidcProviderDispatcher,
    private val jwtProvider: JwtProvider
) {


	@Transactional
	fun signUp(param: UserSignUpParam): JwtResponse {
		val providerId = oidcProviderDispatcher.getProviderId(param.providerType, param.idToken)
		val user = userManageFacade.create(supplyUserBy(param, providerId))

		return supplyJwtResponseBy(jwtProvider.generateToken(user))
	}

	@Transactional(readOnly = true)
	fun signIn(param: UserSignInParam): JwtResponse {
		val providerId = oidcProviderDispatcher.getProviderId(param.providerType, param.idToken)
		val user = userReadFacade.findByProviderId(providerId)

		return supplyJwtResponseBy(jwtProvider.generateToken(user))
	}

	fun signOut(accessToken: String, refreshToken: String) {
		val accessTokenExpiredSecond = jwtProvider.getTokenExpiredSecond(accessToken) - currentTimeMillis()
		val refreshTokenExpiredSecond = jwtProvider.getTokenExpiredSecond(refreshToken) - currentTimeMillis()

		tokenRedisRepository.registerBlackList(accessToken, accessTokenExpiredSecond)
		tokenRedisRepository.registerBlackList(refreshToken, refreshTokenExpiredSecond)
	}

	@Transactional(readOnly = true)
	fun recreateToken(refreshToken: String): JwtResponse {
		val isBlackList = tokenRedisRepository.isBlackListToken(refreshToken)

		if (isBlackList) {
			throw CakkException(ReturnCode.BLACK_LIST_TOKEN)
		}

		val user = jwtProvider.getUser(refreshToken)
		val expiredSecond = jwtProvider.getTokenExpiredSecond(refreshToken) - currentTimeMillis()

		tokenRedisRepository.registerBlackList(refreshToken, expiredSecond)

		return supplyJwtResponseBy(jwtProvider.generateToken(user))
	}
}
